fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+            target (torch.Tensor): target tensor to be used to compute a loss given each models outputs
+
+        Returns:
+            Tuple[TrainingLosses, Dict[str, torch.Tensor]]: Returns relevant loss values from both the global and local
+                model optimization steps. The prediction dictionary contains predictions indexed a "global" and "local"
+                corresponding to predictions from the global and local Ditto models for metric evaluations.
+        """
+
+        # Clear gradients from optimizers if they exist
+        self.optimizers["global"].zero_grad()
+        self.optimizers["local"].zero_grad()
+
+        # Forward pass on both the global and local models
+        preds, features = self.predict(input)
+
+        # Compute all relevant losses
+        losses = self.compute_training_loss(preds, features, target)
+
+        # Take a step with the global model vanilla loss
+        losses.additional_losses["global_loss"].backward()
+        self.optimizers["global"].step()
+
+        # Take a step with the local model using the local loss and Ditto constraint
+        losses.backward["backward"].backward()
+        self.optimizers["local"].step()
+
+        # Return dictionary of predictions where key is used to name respective MetricMeters
+        return losses, preds
+
+    def update_after_step(self, step: int) -> None:
+        if step % self.beta_update_interval == 0:
+            if self.mkmmd_loss_weight and self.init_global_model:
+                # Get the feature distribution of the local and init global features with evaluation mode
+                local_distribution, init_global_distribution = self.update_buffers(self.model, self.init_global_model)
+                # Update betas for the MK-MMD loss based on gathered features during training
+                if self.mkmmd_loss_weight != 0:
+                    self.mkmmd_loss.betas = self.mkmmd_loss.optimize_betas(
+                        X=local_distribution, Y=init_global_distribution, lambda_m=1e-5
+                    )
+                    log(INFO, f"Set optimized betas to minimize distance: {self.mkmmd_loss.betas.squeeze()}.")
+
+        return super().update_after_step(step)
+
+    def update_buffers(
+        self, local_model: torch.nn.Module, init_global_model: torch.nn.Module
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Update the feature buffer of the local and global features."""
+        assert isinstance(local_model, MoonModel)
+        assert isinstance(init_global_model, MoonModel)
+
+        init_state_local_model = local_model.training
+
+        # Set local model to evaluation mode
+        local_model.eval()
+
+        assert not local_model.training
+        assert not init_global_model.training
+
+        local_buffer = []
+        init_global_buffer = []
+
+        with torch.no_grad():
+            for input, target in self.train_loader:
```

Comment:
Super minor, but since we're not using target, we can just not store it `for input, _ in self.train_loader:`
Then we don't have to move it to GPU either.